File: Connector/src/Socket/NetlinkSocket.cpp
#include <netlink/socket.h>
#include <netlink/netlink.h>
#include "../Netlink.hpp"


struct nl_sock* openNetlinkSocket()
{
  struct nl_sock* socket = nl_socket_alloc();
  nl_connect(socket, NETLINK_ROUTE);

  return socket;
}

void closeNetlinkSocket(struct nl_sock* socket)
{
  nl_close(socket);
  nl_socket_free(socket);
}

File: src/dotnetlink/NetlinkSocket/NetlinkSocket.cs
using System;
using libnl;
// ReSharper disable UnusedMember.Global

// ReSharper disable once CheckNamespace
namespace dotnetlink
{
    public unsafe partial class NetlinkSocket
    {
        private nl_sock* _sockFd;

        public NetlinkSocket()
        {
            _sockFd = LibNL3.nl_socket_alloc();
            LibNL3.nl_connect(_sockFd, 0);
        }

        ~NetlinkSocket()
        {
            LibNL3.nl_close(_sockFd);
            LibNL3.nl_socket_free(_sockFd);
            _sockFd = (nl_sock*) 0;
        }
    }
}